packages/luals-annotations/src/annotationBuilder.ts

```diff
@@ -0,0 +1,104 @@
+import { Docs } from "./json-docs";
+import luaLSType from "./luaLSType";
+
+export default class AnnotationBuilder {
+	generatedAnnotations: string[] = [];
+
+	currentClass: string = "";
+	staticMethod: boolean = false;
+
+	constructor() {}
+
+	add(...lines: string[]) {
+		this.generatedAnnotations.push(...lines);
+	}
+
+	addComment(comment: string) {
+		this.add(`--- ${comment}`);
+	}
+
+	addDescription(description: string | string[]) {
+		if (typeof description === "string") {
+			this.addComment(description);
+		} else {
+			description.forEach((line) => this.addComment(line));
+		}
+	}
+
+	addFunction(func: Docs.Function) {
+		const params = func.parameters ?? [];
+		const returns = func.returns ?? [];
+
+		const paramList = params.map(
+			(param) => `---@param ${param.name}${(param.optional && "?") || ""} ${luaLSType(param)}`
+		);
+		const returnList = returns.map((ret) => `---@return ${luaLSType(ret)} ${ret.name ?? ""}`);
+
+		const functionCharacter = this.staticMethod ? "." : ":";
+		const functionPrefix = this.currentClass ? `${this.currentClass}${functionCharacter}` : "";
+
+		const renderFunction = (signature: string) => {
+			this.addDescription(func.description);
+			this.add(...paramList, ...returnList, signature);
+		};
+
+		if (func.aliases) {
+			func.aliases.forEach((alias) => {
+				const signature = `function ${functionPrefix}${alias}(${params.map((param) => param.name)}) end`;
+				renderFunction(signature);
+			});
+		}
+
+		const signature = `function ${functionPrefix}${func.name}(${params.map((param) => param.name)}) end`;
+		renderFunction(signature);
+	}
+
+	addClassDeclaration() {
+		if (!this.currentClass) {
+			throw new Error("Cannot add class declaration without a class");
+		}
+
+		this.add(`${this.currentClass} = {}`);
+	}
+
+	addField(property: Docs.Property) {
+		if (!this.currentClass) {
+			throw new Error("Cannot add field without a class");
+		}
+
+		let description = property.description;
+		if (typeof description !== "string") {
+			// Unfortunately, we can't really have multi-line descriptions for fields
```

Comment:
Might be one to look into for a workaround, or one to raise on the LuaLS github

Reply:
There's not exactly a workaround here, it's just that fields are primitive annotations which means that they're parsed as a single line. Could raise this as an issue on the LuaLS github later.

Reply:
We don't need a workaround even, it's 100% supported by LuaLS

https://github.com/LuaLS/lua-language-server/issues/1531#issuecomment-1239129419

Reply:
Don't know how I missed that?? :trollface: